### PR TITLE
Restructured collaboration.js

### PR DESCRIPTION
**Pull request by hkirat**
Restructured collaboration.js according to https://github.com/mozilla/thimble.mozilla.org/issues/2351#issuecomment-313493101

### CODE REVIEW

**Comment by gideonthomas** [comment]
Fixes mozilla/thimble.mozilla.org#2351

**Comment by hkirat** [comment]
I dont think it fixes  mozilla/thimble.mozilla.org#2351. I'll probably file another PR for that

File: src/editor/Collaboration.js [hunk 1]
```diff
@@ -3,8 +3,18 @@ define(function (require, exports, module) {
...
 
-    function Collaboration() {
-        var webrtc = new SimpleWebRTC({
+    var _webrtc,
+        _pending,
+        _changing,
+        _room,
+        _codemirror;
+
+    function initialize() {
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: let's rename this to `connect`

File: src/editor/Collaboration.js [hunk 2]
```diff
@@ -16,66 +26,64 @@ define(function (require, exports, module) {
...
 
-    Collaboration.prototype.init = function(codemirror) {
-        var self = this;
-        this.webrtc.joinRoom(this.room, function() {
-            self.codemirror = codemirror;
-            self.webrtc.sendToAll("new client", {});
-            self.webrtc.on("createdPeer", function(peer) {
-                self.initializeNewClient(peer);
+    function init(codemirror) {
+        _webrtc.joinRoom(_room, function() {
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: add a check before this:
```js
if(!_webrtc) {
    console.error("WebRTC connection has not been established. Call `connect()` first.");
    return;
}

if(_room) {
    console.warn("Room ", _room, ", already joined");
    return;
}
```

File: src/editor/Collaboration.js [hunk 3]
```diff
@@ -16,66 +26,64 @@ define(function (require, exports, module) {
...
+            _codemirror = codemirror;
+            _webrtc.sendToAll("new client", {});
+            _webrtc.on("createdPeer", function(peer) {
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: ```_webrtc.on("createdPeer", _initializeNewClient);```

File: src/editor/Collaboration.js [hunk 4]
```diff
@@ -16,66 +26,64 @@ define(function (require, exports, module) {
...
+                _initializeNewClient(peer);
             });
 
-            self.webrtc.connection.on('message', function (msg) {
-                self.handleMessage(msg);
+            _webrtc.connection.on('message', function (msg) {
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: `_webrtc.connection.on("message", _handleMessage);`

File: src/editor/Collaboration.js [hunk 5]
```diff
@@ -16,66 +26,64 @@ define(function (require, exports, module) {
...
-    Collaboration.prototype.handleCodemirrorChange = function(delta) {
-        if(this.changing) {
+    function _handleCodemirrorChange (delta) {
+        if(_changing) {
             return;
         }
-        this.changing = true;
-        var cm = this.codemirror;
+        _changing = true;
+        var cm = _codemirror;
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: why create another variable? Just use `_codemirror` below

File: src/editor/Editor.js [hunk 6]
```diff
@@ -103,8 +103,8 @@ define(function (require, exports, module) {
     
 
     var cmOptions         = {};
-    var collabInstance    = new Collaboration();
 
+    Collaboration.initialize();
```
Comment (gideonthomas) [inline src/editor/Editor.js]: blank line after this

File: src/editor/Collaboration.js (hunk 1 above)
Comment (gideonthomas) [inline src/editor/Collaboration.js]: Tbh, I don't think they need to be two separate functions. Can't we just have the `Editor` constructor call a single function called `connect()` which has the contents of both `initialize` and `init`?
```js
function connect(codemirror) {
    if(_webrtc) {
        console.error("Collaboration already initialized");
        return;
    }

    _webrtc = ...
    // Rest of this function

    _webrtc.joinRoom(_room, function() {
        // Rest of the code
    });

    _webrtc.connection.on("message", _handleMessage);
}
```

**Pull request merged**